File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from collections import deque

N, K = map(int, input().split())
INF = int(1e9)
distance = [-1] * 100001
distance[N] = 0

dq = deque()
dq.append(N)

while dq:
    pos = dq.popleft()

    for next_pos, cost in [(pos * 2, 0), (pos - 1, 1), (pos + 1, 1)]:
        if 0 <= next_pos <= 100000:
            if distance[next_pos] == -1 or distance[next_pos] > distance[pos] + cost:
                distance[next_pos] = distance[pos] + cost
                if cost == 0:
                    dq.appendleft(next_pos)
                else:
                    dq.append(next_pos)

print(distance[K])

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.cc
#include <iostream>
#include <deque>
#include <vector>

using namespace std;

const int MAX = 100001;

int main() {
    int N, K;
    cin >> N >> K;

    vector<int> distance(MAX, -1);
    distance[N] = 0;

    deque<int> dq;
    dq.push_back(N);

    while (!dq.empty()) {
        int pos = dq.front();
        dq.pop_front();

        vector<pair<int, int>> moves = {
            {pos * 2, 0},
            {pos - 1, 1},
            {pos + 1, 1}
        };

        for (auto& move : moves) {
            int next_pos = move.first;
            int cost = move.second;

            if (0 <= next_pos && next_pos < MAX) {
                if (distance[next_pos] == -1 || distance[next_pos] > distance[pos] + cost) {
                    distance[next_pos] = distance[pos] + cost;
                    if (cost == 0)
                        dq.push_front(next_pos);
                    else
                        dq.push_back(next_pos);
                }
            }
        }
    }

    cout << distance[K] << '\n';
    return 0;
}
